sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java

```diff
@@ -525,20 +560,38 @@ private void acquireChannel(final Promise<Channel> promise) {
                 for (Channel channel : this.availableChannels) {
 
                     final RntbdRequestManager manager = channel.pipeline().get(RntbdRequestManager.class);
-                    final long pendingRequestCount = manager.pendingRequestCount();
 
-                    if (pendingRequestCount < pendingRequestCountMin) {
-                        pendingRequestCountMin = pendingRequestCount;
-                        candidate = channel;
+                    if (manager == null) {
+                        logger.warn("Channel({} --> {}) closed", channel, this.remoteAddress());
+                    } else {
+                        final long pendingRequestCount = manager.pendingRequestCount();
+
+                        // we accept the risk of reusing the channel even if more than maxPendingRequests are
+                        // queued - by picking the channel with the least number of outstanding requests we load
+                        // balance reasonably
+                        if (isChannelServiceable(channel, true) && pendingRequestCount < pendingRequestCountMin) {
+                            pendingRequestCountMin = pendingRequestCount;
+                            candidate = channel;
+                        }
                     }
                 }
 
-                assert candidate != null;
-
-                this.availableChannels.remove(candidate);
-                doAcquireChannel(promise, candidate);
+                if (candidate != null && this.availableChannels.remove(candidate)) {
```

Comment:
Do we need repeat finding channel if this.availableChannels.remove(candidate)  is false . The way we are going to next channel in else block ?

Reply:
Talked through this offline - current behavior when loadFactor is above 90% is intentional.